Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -12,311 +12,811 @@
-      } else if endLine == lineNumber {
-        endColumn = endLoc.column
-      } else {
-        return nil
+    for annotatedLine in annotatedSourceLines {
+      /// If the current line number is not within the 'rangesToPrint', we continue to the next line
+      /// and set the 'hasLineBeenSkipped' flag, so that we can later insert a marker for the skipped lines.
+      guard rangesToPrint.contains(where: { $0.contains(annotatedLine.sourceLineNumber) }) else {
+        hasLineBeenSkipped = true
+        continue
       }
 
-      if startColumn == endColumn {
-        return nil
+      /// Add a skipped line marker to indicate that lines have been omitted, but only under two conditions:
+      /// 1. Lines have actually been skipped ('hasLineBeenSkipped' is true).
+      /// 2. There's already annotated content in 'annotatedSource', meaning this isn't the start of the annotated lines.
+      /// This ensures that skipped line markers only appear between annotated lines.
+      if hasLineBeenSkipped && !annotatedSource.isEmpty {
+        let skippedLineMarker = diagnosticDecorator.decorateBufferOutline("┆")
+        let skippedLine = indentString + String(repeating: " ", count: maxNumberOfDigits) + " " + skippedLineMarker + "\n"
+        annotatedSource.append(skippedLine)
       }
+      hasLineBeenSkipped = false
 
-      return startColumn..<endColumn
-    }.sorted { (lhs, rhs) in
-      lhs.lowerBound < rhs.lowerBound
-    }.mergingOverlappingRanges()
+      // Append a decorated current source line
+      appendDecoratedSourceLine(
+        to: &annotatedSource,
+        from: annotatedLine,
+        within: tree,
+        via: sourceLocationConverter,
+        indentingWith: indentString,
+        upToMaxDigits: maxNumberOfDigits
+      )
 
-    // Map the column ranges into index ranges within the source string itself.
-    let sourceStringUTF8 = annotatedLine.sourceString.utf8
-    let highlightIndexRanges: [Range<String.Index>] = highlightRanges.map { highlightRange in
-      let startIndex = sourceStringUTF8.index(sourceStringUTF8.startIndex, offsetBy: highlightRange.lowerBound - 1)
-      let endIndex = sourceStringUTF8.index(startIndex, offsetBy: highlightRange.count)
-      return startIndex..<endIndex
-    }
+      // If the line did not end with new line (e.g. the last line), append it manually
+      if annotatedSource.last?.isNewline == false {
+        annotatedSource.append("\n")
+      }
 
-    // Form the annotated string by copying in text from the original source,
-    // highlighting the column ranges.
-    var resultSourceString: String = ""
-    let sourceString = annotatedLine.sourceString
-    var sourceIndex = sourceString.startIndex
-    for highlightRange in highlightIndexRanges {
-      // Text before the highlight range
-      resultSourceString += sourceString[sourceIndex..<highlightRange.lowerBound]
+      // Append all diagnostic messages for the current line
+      appendFormattedDiagnostics(
```

Comment:
Same here, I think would be cleaner as 

```suggestion
      annotatedSource += formattedDiagnostics(
```

Reply:
Yup, it's cleaner. Thanks. Done.